### PR TITLE
Added WaitSynchN/WaitSynch1 tests

### PR DESCRIPTION
**Pull request by bunnei**
When fixing up our implementations of the WaitSynhronizationN and WaitSynchronization1 SVCs, I wrote a ton of tests to make sure I didn't introduce any regressions. These have been sitting around collecting dust for a while because 1) there was a bug that made them not all run sequentially on hardware, and 2) they were messy.. At any rate, I found the solution to 1) when https://github.com/citra-emu/citra/pull/551 was merged (as Citra started hanging like hardware, so I was able to fix the issue), so I'd rather just see these become part of hwtests (to prevent any future regressions) despite the fact that the code isn't super great. These tests exercise a lot of corner cases. At the very least, I spent a bit of time making the code not horrible, but yeah, I don't really plan on spending more time on this in the near term (except maybe adding new tests in the future to help me fix things I know Citra does not do correctly, such as WaitSynch on a Session object).

### CODE REVIEW

File: source/tests/waitsynch/waitsynch.cpp [hunk 1]
```diff
@@ -0,0 +1,1095 @@
...
+u64 thead3_stack[0x1000];
+u64 thead4_stack[0x1000];
+u64 thead5_stack[0x1000];
+u64 thead6_stack[0x1000];
+u64 thead7_stack[0x1000];
+u64 thead8_stack[0x1000];
+u64 thead9_stack[0x1000];
+u64 thead10_stack[0x1000];
+u64 thead11_stack[0x1000];
+u64 thead12_stack[0x1000];
```
Comment (lioncash) [inline source/tests/waitsynch/waitsynch.cpp]: All of the variables from here and up can be made static, can they not?

**Comment by archshift** [comment]
After a cursory glance, LGTM after those variables are made static.

File: source/tests/waitsynch/waitsynch.cpp (hunk 1 above)
Comment (yuriks) [inline source/tests/waitsynch/waitsynch.cpp]: Why are there u64's? This is a 32KB stack for each thread, which seems somewhat excessive.

File: source/tests/waitsynch/waitsynch.cpp [hunk 2]
```diff
@@ -0,0 +1,1095 @@
...
+bool Test_WaitSynchN_01() {
+    s32 output = 0;
+    s32 res = 0;
+
+    svcClearEvent(events[0]);
+    svcClearEvent(events[1]);
+    svcClearEvent(events[2]);
+    svcClearEvent(events[3]);
+
+    svcCreateThread(&theads[3], thread3_handler, 0x0, (u32*)(&thead3_stack[0x1000]), 0x31, 0xfffffffe);
```
Comment (yuriks) [inline source/tests/waitsynch/waitsynch.cpp]: This handle needs to be closed.

File: source/tests/waitsynch/waitsynch.cpp [hunk 3]
```diff
@@ -0,0 +1,1095 @@
...
+    res = svcWaitSynchronizationN(&output, events, 4, 0, U64_MAX);
+    SoftAssert(output == 3);
+    SoftAssert(res == 0);
+
+    svcClearEvent(events[3]);
+
+    svcCreateThread(&theads[0], thread0_handler, 0x0, (u32*)(&thead0_stack[0x1000]), 0x31, 0xfffffffe);
```
Comment (yuriks) [inline source/tests/waitsynch/waitsynch.cpp]: This handle needs to be closed.

File: source/tests/waitsynch/waitsynch.cpp [hunk 4]
```diff
@@ -0,0 +1,1095 @@
...
+
+    svcClearEvent(events[0]);
+    svcClearEvent(events[1]);
+    svcClearEvent(events[2]);
+    svcClearEvent(events[3]);
+
+    svcCreateThread(&theads[0], thread0_handler, 0x0, (u32*)(&thead0_stack[0x1000]), 0x2F, 0xfffffffe);
+    svcCreateThread(&theads[1], thread1_handler, 0x0, (u32*)(&thead1_stack[0x1000]), 0x30, 0xfffffffe);
+    svcCreateThread(&theads[2], thread2_handler, 0x0, (u32*)(&thead2_stack[0x1000]), 0x31, 0xfffffffe);
+    svcCreateThread(&theads[3], thread3_handler, 0x0, (u32*)(&thead3_stack[0x1000]), 0x32, 0xfffffffe);
```
Comment (yuriks) [inline source/tests/waitsynch/waitsynch.cpp]: All these handles needs to be closed.

File: source/tests/waitsynch/waitsynch.cpp [hunk 5]
```diff
@@ -0,0 +1,1095 @@
...
+
+bool Test_WaitSynchN_09() {
+    s32 output = 0;
+    s32 res = 0;
+    Handle handles[] = {events[0], events[1], events[1], events[0], events[1]};
+
+    svcClearEvent(events[0]);
+    svcClearEvent(events[1]);
+
+    svcCreateThread(&theads[1], thread1_handler, 0x0, (u32*)(&thead1_stack[0x1000]), 0x31, 0xfffffffe);
```
Comment (yuriks) [inline source/tests/waitsynch/waitsynch.cpp]: This handle needs to be closed.

File: source/tests/waitsynch/waitsynch.cpp [hunk 6]
```diff
@@ -0,0 +1,1095 @@
...
+#include "output.h"
+#include "common/string_funcs.h"
+#include "common/scope_exit.h"
+#include "tests/test.h"
+#include "tests/waitsynch/waitsynch.h"
+
+namespace WaitSynch {
+
+Handle events[4];
+Handle theads[13];
```
Comment (yuriks) [inline source/tests/waitsynch/waitsynch.cpp]: Typo: `theads`

File: source/tests/waitsynch/waitsynch.cpp (hunk 1 above)
Comment (yuriks) [inline source/tests/waitsynch/waitsynch.cpp]: Typo: `thead`

File: source/tests/waitsynch/waitsynch.cpp [hunk 7]
```diff
@@ -0,0 +1,1095 @@
...
+
+    return true;
+}
+
+// Test WaitSynchN with svcDuplicateHandle (one event signaled, with wait)
+
+bool Test_WaitSynchN_10() {
+    s32 output = 0;
+    s32 res = 0;
+    Handle duplicate0, duplicate1;
```
Comment (yuriks) [inline source/tests/waitsynch/waitsynch.cpp]: Handles needs to be closed.

File: source/tests/waitsynch/waitsynch.cpp [hunk 8]
```diff
@@ -0,0 +1,1095 @@
...
+
+// Test WaitSynchN with svcDuplicateHandle (one event signaled, with wait)
+
+bool Test_WaitSynchN_11() {
+    s32 output = 0;
+    s32 res = 0;
+    Handle duplicate0, duplicate1;
+
+    svcDuplicateHandle(&duplicate0, events[1]);
+    svcDuplicateHandle(&duplicate1, events[1]);
```
Comment (yuriks) [inline source/tests/waitsynch/waitsynch.cpp]: Handles needs to be closed.

File: source/tests/waitsynch/waitsynch.cpp [hunk 9]
```diff
@@ -0,0 +1,1095 @@
...
+}
+
+bool Test_WaitSynchN_15() {
+    s32 output = 0;
+    s32 res = 0;
+
+    svcClearEvent(events[0]);
+    svcClearEvent(events[1]);
+
+    svcCreateThread(&theads[4], thread4_handler, 0x0, (u32*)(&thead4_stack[0x1000]), 0x31, 0xfffffffe);
```
Comment (yuriks) [inline source/tests/waitsynch/waitsynch.cpp]: Handle needs to be closed.

File: source/tests/waitsynch/waitsynch.cpp [hunk 10]
```diff
@@ -0,0 +1,1095 @@
...
+    s32 output = 0;
+    s32 res = 0;
+
+    svcReleaseMutex(mutexes[0]);
+    svcReleaseMutex(mutexes[1]);
+
+    svcClearEvent(events[0]);
+    svcClearEvent(events[1]);
+
+    svcCreateThread(&theads[5], thread5_handler, 0x0, (u32*)(&thead5_stack[0x1000]), 0x31, 0xfffffffe);
```
Comment (yuriks) [inline source/tests/waitsynch/waitsynch.cpp]: Handles needs to be closed.

File: source/tests/waitsynch/waitsynch.cpp [hunk 11]
```diff
@@ -0,0 +1,1095 @@
...
+    svcReleaseMutex(mutexes[0]);
+    svcReleaseMutex(mutexes[0]);
+
+    svcReleaseMutex(mutexes[1]);
+    svcReleaseMutex(mutexes[1]);
+
+    res = svcReleaseMutex(mutexes[0]);
+    SoftAssert(res == 0);
+
+    svcCreateThread(&theads[5], thread5b_handler, 0x0, (u32*)(&thead5_stack[0x1000]), 0x31, 0xfffffffe);
```
Comment (yuriks) [inline source/tests/waitsynch/waitsynch.cpp]: Handles needs to be closed.

File: source/tests/waitsynch/waitsynch.cpp [hunk 12]
```diff
@@ -0,0 +1,1095 @@
...
+    s32 output = 0;
+    s32 res = 0;
+
+    svcReleaseMutex(mutexes[0]);
+    svcReleaseMutex(mutexes[1]);
+
+    svcClearEvent(events[0]);
+    svcClearEvent(events[1]);
+
+    svcCreateThread(&theads[6], thread6_handler, 0x0, (u32*)(&thead6_stack[0x1000]), 0x31, 0xfffffffe);
```
Comment (yuriks) [inline source/tests/waitsynch/waitsynch.cpp]: Handles needs to be closed.

File: source/tests/waitsynch/waitsynch.cpp [hunk 13]
```diff
@@ -0,0 +1,1095 @@
...
+    s32 output = 0;
+    s32 res = 0;
+    Handle handles[] = { events[2], mutexes[0] };
+
+    svcReleaseMutex(mutexes[0]);
+    svcClearEvent(events[0]);
+    svcClearEvent(events[1]);
+    svcSignalEvent(events[2]);
+
+    svcCreateThread(&theads[7], thread7_handler, 0x0, (u32*)(&thead7_stack[0x1000]), 0x31, 0xfffffffe);
```
Comment (yuriks) [inline source/tests/waitsynch/waitsynch.cpp]: Handles needs to be closed.

File: source/tests/waitsynch/waitsynch.cpp [hunk 14]
```diff
@@ -0,0 +1,1095 @@
...
+    s32 output = 0;
+    s32 res = 0;
+    Handle handles[] = { events[2], mutexes[0] };
+
+    svcReleaseMutex(mutexes[0]);
+    svcClearEvent(events[0]);
+    svcClearEvent(events[1]);
+    svcSignalEvent(events[2]);
+
+    svcCreateThread(&theads[8], thread8_handler, 0x0, (u32*)(&thead8_stack[0x1000]), 0x31, 0xfffffffe);
```
Comment (yuriks) [inline source/tests/waitsynch/waitsynch.cpp]: Handles needs to be closed.

File: source/tests/waitsynch/waitsynch.cpp [hunk 15]
```diff
@@ -0,0 +1,1095 @@
...
+}
+
+bool Test_WaitSynchN_22() {
+    s32 output = 0;
+    s32 res = 0;
+
+    svcClearEvent(events[0]);
+    svcSignalEvent(events[1]);
+
+    svcCreateThread(&theads[9], thread9_handler, 0x0, (u32*)(&thead9_stack[0x1000]), 0x31, 0xfffffffe);
```
Comment (yuriks) [inline source/tests/waitsynch/waitsynch.cpp]: Handles needs to be closed.

File: source/tests/waitsynch/waitsynch.cpp [hunk 16]
```diff
@@ -0,0 +1,1095 @@
...
+
+    Handle service = 0;
+    srvGetServiceHandle(&service, "APT:U");
+
+    // Make sure we don't timeout
+    res = svcWaitSynchronizationN(&output, &service, 1, 0, ONE_SECOND/2);
+    SoftAssert(output == -1);
+    SoftAssert(res == 0x09401BFE);
+    
+    svcCloseHandle(service);
```
Comment (yuriks) [inline source/tests/waitsynch/waitsynch.cpp]: Should use `SCOPE_EXIT` to close this, in case the asserts fail.

File: source/tests/waitsynch/waitsynch.cpp [hunk 17]
```diff
@@ -0,0 +1,1095 @@
...
+bool Test_WaitSynchN_25() {
+    s32 output = 0;
+    s32 res = 0;
+
+    svcReleaseMutex(mutexes[0]);
+
+    svcClearEvent(events[0]);
+    svcClearEvent(events[1]);
+
+    svcCreateThread(&theads[10], thread10_handler, 0x0, (u32*)(&thead10_stack[0x1000]), 0x31, 0xfffffffe);
```
Comment (yuriks) [inline source/tests/waitsynch/waitsynch.cpp]: Handles needs to be closed.

File: source/tests/waitsynch/waitsynch.cpp [hunk 18]
```diff
@@ -0,0 +1,1095 @@
...
+
+// Test WaitSynch1 with wait=1
+
+bool Test_WaitSynch1_01() {
+    s32 output = 0;
+    s32 res = 0;
+
+    svcClearEvent(events[0]);
+
+    svcCreateThread(&theads[0], thread0_handler, 0x0, (u32*)(&thead0_stack[0x1000]), 0x31, 0xfffffffe);
```
Comment (yuriks) [inline source/tests/waitsynch/waitsynch.cpp]: Handles needs to be closed.

File: source/tests/waitsynch/waitsynch.cpp [hunk 19]
```diff
@@ -0,0 +1,1095 @@
...
+    return true;
+}
+
+// Test WaitSynch1 with a service handle
+
+bool Test_WaitSynch1_03() {
+    s32 res = 0;
+    Handle service = 0;
+
+    srvGetServiceHandle(&service, "soc:U");
```
Comment (yuriks) [inline source/tests/waitsynch/waitsynch.cpp]: Handles needs to be closed.

File: source/tests/waitsynch/waitsynch.cpp [hunk 20]
```diff
@@ -0,0 +1,1095 @@
...
+bool Test_WaitSynch1_04() {
+    s32 output = 0;
+    s32 res = 0;
+
+    svcReleaseMutex(mutexes[0]);
+
+    svcClearEvent(events[0]);
+    svcClearEvent(events[1]);
+
+    svcCreateThread(&theads[11], thread11_handler, 0x0, (u32*)(&thead11_stack[0x1000]), 0x31, 0xfffffffe);
```
Comment (yuriks) [inline source/tests/waitsynch/waitsynch.cpp]: Handles needs to be closed.

File: source/tests/waitsynch/waitsynch.cpp [hunk 21]
```diff
@@ -0,0 +1,1095 @@
...
+    s32 output = 0;
+    s32 res = 0;
+
+    svcReleaseMutex(mutexes[0]);
+
+    svcClearEvent(events[0]);
+    svcClearEvent(events[1]);
+    svcClearEvent(events[2]);
+
+    svcCreateThread(&theads[12], thread12_handler, 0x0, (u32*)(&thead12_stack[0x1000]), 0x31, 0xfffffffe);
```
Comment (yuriks) [inline source/tests/waitsynch/waitsynch.cpp]: Handles needs to be closed.

File: source/tests/waitsynch/waitsynch.cpp [hunk 22]
```diff
@@ -0,0 +1,1095 @@
...
+    res = svcWaitSynchronization(shared_mem, ONE_SECOND/2);
+    SoftAssert(res == 0xD8E007F7);
+
+    res = svcWaitSynchronizationN(&output, &shared_mem, 1, 0, ONE_SECOND/2);
+    SoftAssert(res == 0xD8E007F7);
+
+    res = svcWaitSynchronizationN(&output, &shared_mem, 1, 1, ONE_SECOND/2);
+    SoftAssert(res == 0xD8E007F7);
+
+    svcCloseHandle(service);
```
Comment (yuriks) [inline source/tests/waitsynch/waitsynch.cpp]: Use `SCOPE_EXIT`

File: source/tests/waitsynch/waitsynch.cpp [hunk 23]
```diff
@@ -0,0 +1,1095 @@
...
+    
+    // res[1] = svcWaitSynchronizationN(&output[1], &service, 1, 0, ONE_SECOND/2);
+    // SoftAssert(output[1] == -1);
+    // SoftAssert(res[1] == 0x09401BFE);
+    
+    res[2] = svcWaitSynchronizationN(&output[2], &service, 1, 1, ONE_SECOND/2);
+    // SoftAssert(output[2] == -1);
+    // SoftAssert(res[2] == 0); // Wrong in citra
+    
+    svcCloseHandle(service);
```
Comment (yuriks) [inline source/tests/waitsynch/waitsynch.cpp]: `SCOPE_EXIT`

File: source/tests/waitsynch/waitsynch.cpp [hunk 24]
```diff
@@ -0,0 +1,1095 @@
...
+
+    return true;
+}
+
+// Test WaitSynch with an AddressArbiter
+
+bool Test_WaitSynch_KernelType_AddressArbiter() {
+    s32 output;
+    s32 res;
+    Handle arb = 0;
```
Comment (yuriks) [inline source/tests/waitsynch/waitsynch.cpp]: Handles needs to be closed.

**Comment by yuriks** [comment]
Mostly a butt-load of handle leaks. If you don't feel like dealing with them I can fix it up.

File: source/tests/waitsynch/waitsynch.cpp (hunk 1 above)
Comment (Subv) [inline source/tests/waitsynch/waitsynch.cpp]: Any reason to not use a two-dimensional array here?

File: source/tests/waitsynch/waitsynch.cpp [hunk 25]
```diff
@@ -0,0 +1,1133 @@
...
+#include "tests/waitsynch/waitsynch.h"
+
+namespace WaitSynch {
+
+static Handle events[4];
+static Handle threads[13];
+static Handle mutexes[4];
+static Handle semaphore;
+
+static u32 thead0_stack[0x1000];
```
Comment (yuriks) [inline source/tests/waitsynch/waitsynch.cpp]: typos: `thead`

File: source/tests/waitsynch/waitsynch.cpp [hunk 26]
```diff
@@ -0,0 +1,1133 @@
...
+
+    Handle service = 0;
+    SoftAssert(srvGetServiceHandle(&service, "APT:U") == 0);
+    SoftAssert(service != 0);
+    SCOPE_EXIT({ svcCloseHandle(service); });
+
+    // Make sure we don't timeout
+    res = svcWaitSynchronizationN(&output, &service, 1, 0, ONE_SECOND/2);
+    SoftAssert(output == -1);
+    SoftAssert(res == 0x09401BFE);
```
Comment (yuriks) [inline source/tests/waitsynch/waitsynch.cpp]: Can you name/decode these result codes? https://gist.github.com/yuriks/715d3478168eb69a28fa

File: source/tests/waitsynch/waitsynch.cpp [hunk 27]
```diff
@@ -0,0 +1,1133 @@
...
+
+bool Test_WaitSynchN_24() {
+    s32 output = 0;
+    s32 res = 0;
+    Handle handles[] = { events[0], 0xDEADBEEF };
+
+    svcSignalEvent(events[0]);
+
+    res = svcWaitSynchronizationN(&output, handles, 2, 0, U64_MAX);
+    SoftAssert(res == 0xD8E007F7);
```
Comment (yuriks) [inline source/tests/waitsynch/waitsynch.cpp]: Name this

File: source/tests/waitsynch/waitsynch.cpp [hunk 28]
```diff
@@ -0,0 +1,1133 @@
...
+
+    res = svcWaitSynchronizationN(&output, handles, 2, 1, ONE_SECOND / 2);
+    SoftAssert(res == 0xD8E007F7);
```
Comment (yuriks) [inline source/tests/waitsynch/waitsynch.cpp]: Name this

File: source/tests/waitsynch/waitsynch.cpp [hunk 29]
```diff
@@ -0,0 +1,1133 @@
...
+// Test WaitSynch1 with a timeout and with wait=0
+
+bool Test_WaitSynch1_00() {
+    s32 output = 0;
+    s32 res = 0;
+
+    svcClearEvent(events[0]);
+
+    res = svcWaitSynchronization(events[0], ONE_SECOND / 2);
+    SoftAssert(res == 0x09401BFE);
```
Comment (yuriks) [inline source/tests/waitsynch/waitsynch.cpp]: Name this

File: source/tests/waitsynch/waitsynch.cpp [hunk 30]
```diff
@@ -0,0 +1,1133 @@
...
+    SoftAssert(res == 0);
+
+    return true;
+}
+
+// Test WaitSynch1 with an invalid handle
+
+bool Test_WaitSynch1_02() {
+    s32 res = svcWaitSynchronization(0xDEADBEEF, U64_MAX);
+    SoftAssert(res == 0xD8E007F7);
```
Comment (yuriks) [inline source/tests/waitsynch/waitsynch.cpp]: Name this

File: source/tests/waitsynch/waitsynch.cpp [hunk 31]
```diff
@@ -0,0 +1,1133 @@
...
+
+bool Test_WaitSynch1_03() {
+    s32 res = 0;
+    Handle service = 0;
+
+    srvGetServiceHandle(&service, "soc:U");
+    SCOPE_EXIT({ svcCloseHandle(service); });
+
+    res = svcWaitSynchronization(service, ONE_SECOND/2);
+    SoftAssert(res == 0x09401BFE);
```
Comment (yuriks) [inline source/tests/waitsynch/waitsynch.cpp]: Name this

File: source/tests/waitsynch/waitsynch.cpp [hunk 32]
```diff
@@ -0,0 +1,1133 @@
...
+    // Get an example of a shared_mem handle
+    u32* cmdbuf=getThreadCommandBuffer();
+    cmdbuf[0] = 0xa0000;
+    svcSendSyncRequest(service);
+    Handle shared_mem = cmdbuf[3];
+    SoftAssert(shared_mem != 0);
+
+    // Test WaitSynch with a shared mem handle
+    res = svcWaitSynchronization(shared_mem, ONE_SECOND/2);
+    SoftAssert(res == 0xD8E007F7);
```
Comment (yuriks) [inline source/tests/waitsynch/waitsynch.cpp]: Name this

File: source/tests/kernel/waitsynch.cpp [hunk 33]
```diff
@@ -0,0 +1,1137 @@
...
+
+// Test WaitSynchN with invalid handle count
+
+bool Test_WaitSynchN_14() {
+    s32 output = 0;
+    s32 res = 0;
+
+    res = svcWaitSynchronizationN(&output, events, -1, 0, U64_MAX);
+    SoftAssert(output == 0);
+    SoftAssert(res == 0xE0E01BFD);
```
Comment (yuriks) [inline source/tests/kernel/waitsynch.cpp]: Name this :V

File: source/tests/kernel/waitsynch.cpp [hunk 34]
```diff
@@ -0,0 +1,1137 @@
...
+
+// Test WaitSynchN with invalid handles pointer
+
+bool Test_WaitSynchN_16() {
+    s32 output = 0;
+    s32 res = 0;
+
+    res = svcWaitSynchronizationN(&output, nullptr, 0, 0, ONE_SECOND / 2);
+    //SoftAssert(output == 0); // TODO: Figure out how output is set, same as Test_WaitSynchN_02
+    SoftAssert(res == 0xD8E007F6);
```
Comment (yuriks) [inline source/tests/kernel/waitsynch.cpp]: Name these

**Comment by yuriks** [comment]
All cleared for landing. (After squashing, of course.)

**Comment by yuriks** [comment]
:rocket:
